core/touch_gesture.js

```diff
@@ -0,0 +1,261 @@
+  this.onMoveWrapper_ = Blockly.bindEventWithChecks_(
+    document, 'mousemove', null, this.handleMove.bind(this), /*opt_noCaptureIdentifier*/ true);
+  this.onUpWrapper_ = Blockly.bindEventWithChecks_(
+    document, 'mouseup', null, this.handleUp.bind(this), /*opt_noCaptureIdentifier*/ true);
+};
+
+/**
+ * Handle a mouse down, touch move, or pointer move event.
+ * @param {!Event} e A mouse move, touch move, or pointer move event.
+ * @package
+ */
+Blockly.TouchGesture.prototype.handleStart = function(e) {
+  if (Blockly.Touch.isTouchEvent(e) && !this.isDragging()) {
+    this.handleTouchStart(e);
+
+    if (this.isMultiTouch()) {
+      Blockly.longStop_();
+    }
+  }
+};
+
+/**
+ * Handle a mouse move, touch move, or pointer move event.
+ * @param {!Event} e A mouse move, touch move, or pointer move event.
+ * @package
+ */
+Blockly.TouchGesture.prototype.handleMove = function(e) {
+  if (Blockly.Touch.isTouchEvent(e) && !this.isDragging()) {
+    this.handleTouchMove(e);
+  }
+  if (!this.isMultiTouch() && Blockly.Touch.shouldHandleEvent(e)) {
+    Blockly.TouchGesture.superClass_.handleMove.call(this, e);
+  }
+};
+
+/**
+ * Handle a mouse up, touch end, or pointer up event.
+ * @param {!Event} e A mouse up, touch end, or pointer up event.
+ * @package
+ */
+Blockly.TouchGesture.prototype.handleUp = function(e) {
+  if (Blockly.Touch.isTouchEvent(e) && !this.isDragging()) {
+    this.handleTouchEnd(e);
+  }
+  if (!this.isMultiTouch() || this.isDragging()) {
+    if (!Blockly.Touch.shouldHandleEvent(e)) {
+      return;
+    }
+    Blockly.TouchGesture.superClass_.handleUp.call(this, e);
+  } else {
+    e.preventDefault();
+    e.stopPropagation();
+
+    this.dispose();
+  }
+};
+
+/**
+ * Whether this gesture is part of a mulit-touch gesture.
+ * @return {boolean} whether this gesture was a click on a workspace.
+ * @package
+ */
+Blockly.TouchGesture.prototype.isMultiTouch = function() {
+  return this.isMultiTouch_;
+};
+
+/**
+ * Sever all links from this object.
+ * @package
+ */
+Blockly.TouchGesture.prototype.dispose = function() {
+  Blockly.TouchGesture.superClass_.dispose.call(this);
+
+  if (this.onStartWrapper_) {
+    Blockly.unbindEvent_(this.onStartWrapper_);
+  }
+  if (this.onGestureChangeWrapper_) {
+    Blockly.unbindEvent_(this.onGestureChangeWrapper_);
+  }
+  if (this.onGestureEndWrapper_) {
+    Blockly.unbindEvent_(this.onGestureEndWrapper_);
+  }
+};
+
+/**
+ * Handle a touch start or pointer down event and keep track of current pointers.
+ * @param {!Event} e A touch start, or pointer down event.
+ * @package
+ */
+Blockly.TouchGesture.prototype.handleTouchStart = function(e) {
+  var pointerId = Blockly.Touch.getTouchIdentifierFromEvent(e);
+  // store the pointerId in the current list of pointers
+  this.cachedPoints[pointerId] = this.getTouchPoint(e);
+  var pointers = Object.keys(this.cachedPoints);
+  // If two pointers are down, check for pinch gestures
```

Comment:
Yes, if two pointers are down, we assume you are trying to zoom. In the future, if there are more multi-touch scenarios supporter this can be altered.